app/src/main/java/org/simple/clinic/user/UserSession.kt

```diff
@@ -74,6 +84,39 @@ class UserSession @Inject constructor(
         }
   }
 
+  fun loginFromOngoingRegistrationEntry(): Completable {
+    return ongoingRegistrationEntry()
+        .flatMap { entry ->
+          passwordHasher.hash(entry.pin!!)
+              .map { passwordDigest ->
+                LoggedInUser(
+                    uuid = entry.uuid!!,
+                    fullName = entry.fullName!!,
+                    phoneNumber = entry.phoneNumber!!,
+                    pinDigest = passwordDigest,
+                    facilityUuid = UUID.fromString("38b16571-4105-41ae-b8dd-8e3312cdb96c"),
+                    createdAt = entry.createdAt!!,
+                    updatedAt = entry.createdAt,
+                    status = LoggedInUser.Status.WAITING_FOR_APPROVAL
+                )
+              }
+        }
+        .flatMapCompletable { storeUser(it) }
+  }
+
+  fun register(): Single<RegistrationResult> {
+    return loggedInUser()
+        .map { (user) -> user }
+        .firstOrError()
+        .map(::RegistrationRequest)
+        .flatMap { registrationApi.createUser(it) }
+        .flatMap<RegistrationResult> {
+          storeUser(it.loggedInUser)
+              .andThen(Single.just(RegistrationResult.Success()))
+        }
+        .onErrorReturn { RegistrationResult.Error() }
```

Comment:
How do we handle the case where the server fails for some reason (Unreachable/500 error) which means the server might be able to handle the request if attempted again after a while?

The current implementation will not retry the request in these cases and will never attempt to log the user back in again. We need to separate errors that we can't recover from (say missing facility) from the errors that we can recover from. @saket @govindkrjoshi Thoughts?

The updated flow for self registration converts this call to a blocking one, so I'm guessing this is acceptable for now, but we need to figure out how to handle these cases.